cratetorrent/src/disk.rs

```diff
@@ -365,14 +367,129 @@ mod tests {
             torrent_disk_alert_port.recv().await
         {
             // piece is complete so it should be hashed but be invalid
-            assert!(matches!(batch.is_piece_valid, Some(false)));
+            assert_eq!(batch.is_piece_valid, Some(false));
             // verify that the message doesn't contain any blocks
             assert!(batch.blocks.is_empty());
         } else {
             assert!(false, "piece could not be written to disk");
         }
     }
 
+    /// Tests writing of a complete valid torrent's pieces and verifying that an
+    /// alert of each disk write is returned by the disk task.
+    #[tokio::test]
+    async fn should_read_piece_blocks() {
+        let (_, disk_handle, mut alert_port) = spawn().unwrap();
+
+        let Env {
+            id,
+            pieces,
+            piece_hashes,
+            info,
+        } = Env::new("read_piece_blocks");
+
+        // allocate torrent via channel
+        disk_handle
+            .allocate_new_torrent(id, info.clone(), piece_hashes)
+            .unwrap();
+
+        // wait for result on alert port
+        let mut torrent_disk_alert_port =
+            if let Some(Alert::TorrentAllocation(Ok(allocation))) =
+                alert_port.recv().await
+            {
+                allocation.alert_port
+            } else {
+                assert!(false, "torrent could not be allocated");
+                return;
+            };
+
+        // write second piece to disk
+        let index = 1;
+        let piece = &pieces[index];
+        for_each_block(index, piece.len() as u32, |block| {
+            let block_end = block.offset + block.len;
+            let data = &piece[block.offset as usize..block_end as usize];
+            debug_assert_eq!(data.len(), block.len as usize);
+            println!("Writing piece {} block {:?}", index, block);
+            disk_handle.write_block(id, block, data.to_vec()).unwrap();
+        });
+
+        // wait for disk write result
+        assert!(torrent_disk_alert_port.recv().await.is_some());
+
+        // set up channels to communicate disk read results
+        let (chan, mut port) = mpsc::unbounded_channel();
+
+        // read each block in piece
+        let block_count = block_count(piece.len() as u32) as u32;
```

Comment:
Can we clean this up? The issue was that async closures are not stable yet so there was no way to call `port.recv().await` inside the closure passed to `for_each_block`, so it had to be manually inlined here.

Reply:
This is probably not that important for now, let's not block this MR